StandardLibraryEssentials/StepForwardWithSTL/LeaderboardCreation/src/leaderboard.cpp

```diff
@@ -0,0 +1,115 @@
+#include "leaderboard.hpp"
+
+void Leaderboard::retrieveScores() {
+    std::ifstream file(filepath);
+    if (file.is_open()) {
+        std::string line;
+        while (getline(file, line)) {
+            std::stringstream ss(line);
+            std::string name;
+            unsigned int score;
+            getline(ss, name, ',');
+            ss >> score;
+            this->scores.insert({score, name});
+        }
+        file.close();
+    } else {
+        std::cout << "Unable to open file: " << this->filepath << std::endl;
+    }
+}
+
+void Leaderboard::addScore(unsigned int score, std::string name) {
+    scores.insert({score, name});
+    if (scores.size() == 11) {
+        scores.erase(std::prev(scores.end()));
+    }
+}
+
+void Leaderboard::saveScores() {
+    std::ofstream
+            file(filepath);
+    if (file.is_open()) {
+        for (const auto &[score, name]: scores) {
+            file << name << "," << score << std::endl;
+        }
+        file.close();
+    } else {
+        std::cout << "Unable to open file: " << this->filepath << std::endl;
+    }
+}
+
+void Leaderboard::updateScore(unsigned int score) {
+    retrieveScores();
+    addScore(score);
+    saveScores();
+}
+
+sf::Text createText(const std::string &str, const sf::Font &font, const unsigned int size, const float x, const float y) {
+    sf::Text text;
+    text.setFont(font);
+    text.setCharacterSize(size);
+    text.setFillColor(sf::Color::White);
+    text.setString(str);
+    text.setPosition(x, y);
+    return text;
+}
+
+void Leaderboard::drawLeaderboard(sf::RenderWindow &window, const TextureManager &textureManager, ScoreMap scores) {
+    // show leaderboard frame
+    sf::Sprite frame;
+    frame.setTexture(*textureManager.getTexture(GameTextureID::LEADERBOARD));
+    frame.setPosition(100, 75);
+    window.draw(frame);
+
+    sf::Font font;
+    font.loadFromFile("resources/pixelLetters.ttf");
+
+    sf::Text header = createText("LEADERBOARD", font, 40, 300, 80);
+    window.draw(header);
+
+    sf::Text posHeader = createText("POS", font, 26, 157, 162);
```

Comment:
Why do you use such specific constants like `157`? 
Can it be a round number at least, like  `160` for example?

Reply:
Current leaderboard version is pixel-perfectly fitted, and changing any of these numbers would lead to visible inaccuracy. These numbers depend on a font, so I guess there is no clear way to make them look better.
We can move frame to make those numbers more round, but current frame position is in a perfect center of the window.

Reply:
I am afraid pixel-fitting is not a very reliable solution : (

Could you remind what was the problem with drawing the leaderboard table using the shapes drawing capabilities of SFML?
https://www.sfml-dev.org/tutorials/2.6/graphics-shape.php

I think we should re-evaluate this solution. 
With the solution when leaderboard is drawn as a number of shapes (rectangles and lines), you have more control over coordinates, and you won't have to pixel-fit the text to the texture image used to emulate leaderboard table.

Reply:
The approach with a single table in the form of a png file was chosen because it greatly boosted the process of creating a table during the development (it wasn't necessary to match the edges of 33 rectangles or 16 lines).
 
By redesigning the table using the tools provided by SFML, we will not be able to avoid all the same problems with pixel-by-pixel adjustment of text fields, since all this comes from a specific font and a specific size. When I made a prototype with JetBrainsMono, I redid all the text fields in height, because even at the same size they were displayed completely differently. Anyway, the screen will display the same lines in the same places, if we want to keep leaderboard in the same place and size.